File: IrrgartenJava/src/modelo/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package modelo;

/**
 *
 * @author nmatl
 */
public class Labyrinth {
    static private final char BLOCK_CHAR = 'X';
   static private final char EMPTY_CHAR = ' ';
   static private final char MONSTER_CHAR = 'M';
   static private final char COMBAT_CHAR = 'C';
   static private final char EXIT_CHAR = 'E';
   static private final int ROW = 0;
   static private final int COL = 1;
   
   private int nRows;
   private int nCols;
   private int exitRow;
   private int exitCol;
   
   private Monster[][] monsters;
   private Player[][] players;
   private char[][] labyrinth;
   
   
   public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
       this.nRows = nRows;
       this.nCols = nCols;
       this.exitRow = exitRow;
       this.exitCol = exitCol;
       
       this.labyrinth = new char [nRows][nCols];
       this.monsters = new Monster [nRows][nCols];
       this.players = new Player [nRows][nCols];
       for(int i = 0; i < nRows; i++){
           for(int j = 0; j < nCols; j++){
            this.labyrinth[i][j] = Labyrinth.EMPTY_CHAR;
            this.players[i][j] = null;
            this.monsters[i][j] = null;
           }
       }
       this.labyrinth[exitRow][exitCol] = Labyrinth.EXIT_CHAR;
       
       
   }
   
   public boolean haveAWinner(){
       return players[exitRow][exitCol] != null;
   }
  
   
   @Override
   public String toString() {
        String result = "";

        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {

                if (players[i][j] != null) {
                    result += 'P';
                } else if (monsters[i][j] != null) {
                    result += MONSTER_CHAR;
                } else {
                    result += labyrinth[i][j];
                }
            }
            result += '\n';
        }

        return result;
    }
   
    private boolean posOk(int row, int col) {
        return (row >= 0 && row < nRows && col >= 0 
                && col < nCols);
    }
    
    private boolean emptyPos(int row, int col){
        if(posOk(row,col)){
            return labyrinth[row][col] == EMPTY_CHAR;
        }
        return false;
    }
    
     private boolean monsterPos(int row, int col){
        if(posOk(row,col)){
            return labyrinth[row][col] == MONSTER_CHAR;
        }
        return false;
    }
    
      private boolean exitPos(int row, int col){
        if(posOk(row,col)){
            return labyrinth[row][col] == EXIT_CHAR;
        }
        return false;
    }
     private boolean combatPos(int row, int col) {
        return posOk(row, col) && labyrinth[row][col] == COMBAT_CHAR;
    }

    private boolean canStepOn(int row, int col) {
        return posOk(row, col) && labyrinth[row][col] != BLOCK_CHAR;
    }
    
    private void updateOldPos(int row, int col) {
        if (posOk(row, col) && combatPos(row, col)) {
            if (row == exitRow && col == exitCol) {
                labyrinth[row][col] = EXIT_CHAR;
            } else if (monsters[row][col] != null) {
                labyrinth[row][col] = MONSTER_CHAR;
            } else {
                labyrinth[row][col] = EMPTY_CHAR;
            }
        }
    }
      
     private int[] dir2pos(int row, int col, Directions direction) {
          int[] newPos = {row,col};
          switch(direction){
                  case UP:
                    newPos[ROW]--;
                    break;
                  case DOWN:
                      newPos[ROW]++;
                      break;
                  case LEFT:
                      newPos[COL]--;
                      break;
                  case RIGHT:
                      newPos[COL]++;
                      break;
          }  
          return newPos;       
     }
     
     private int[] randomEmptyPos(){
         while (true) {
            int row = Dice.randomPos(nRows);
            int col = Dice.randomPos(nCols);
            if (emptyPos(row, col)) {
                return new int[]{row, col};
            }
        }
     
     }
   
}

File: IrrgartenRuby/modelo/labyrinth.rb
# frozen_string_literal: true
require_relative 'directions'
require_relative 'dice'
module Modelo
  class Labyrinth
    @@BLOCK_CHAR = 'X'
    @@EMPTY_CHAR = '-'
    @@MONSTER_CHAR = 'M'
    @@COMBAT_CHAR= 'C'
    @@EXIT_CHAR = 'E'
    @@ROW = 0 #Su objetivo es que no haya números mágicos
    @@COL = 1

    def initialize(n_rows,n_cols, exit_row, exit_col)
      @n_rows=n_rows
      @n_cols=n_cols
      @exit_row=exit_row
      @exit_col=exit_col
      @monsters = Array.new(n_rows) {Array.new(n_cols, nil)} #nil es una inicialización a puntero (nullptr)
      @players = Array.new(n_rows) {Array.new(n_cols, nil)}
      @labyrinth = Array.new(n_rows) {Array.new(n_cols, @@EMPTY_CHAR)}
      @labyrinth[exit_row][exit_col] = @@EXIT_CHAR
    end

    def have_a_winner?
       @players[@exit_row][@exit_col] != nil
    end

    def to_s
      result = ""

      for i in 0...@n_rows
        for j in 0...@n_cols

          if @players[i][j] != nil
            result += 'P'
          elsif @monsters[i][j] != nil
            result += 'M'
          else
            result += @labyrinth[i][j]
          end

        end
        result += "\n"
      end

      result
    end

    private

    def pos_ok?(row, col)
      row >= 0 && row <= @n_rows && col >= 0 && col <= @n_col
    end

    def empty_pos?(row,col)
      if pos_ok?(row,col)
        return @labyrinth[row][col] == @@EMPTY_CHAR
      end
    end

    def monster_pos?(row,col)
      if pos_ok?(row,col)
        return @labyrinth[row][col] == @@MONSTER_CHAR
      end
    end

    def exist_pos?(row,col)
      if pos_ok?(row,col)
        return @labyrinth[row][col] == @@EXIT_CHAR
      end
    end

    def combat_pos?(row,col)
      if pos_ok?(row,col)
        return @labyrinth[row][col] == @@COMBAT_CHAR
      end
    end

    def can_step_on?(row,col)
      pos_ok?(row,col) && @labyrinth[row][col] != @@BLOCK_CHAR
    end

    def update_old_pos(row,col)
      if pos_ok?(row,col) && combat_pos?(row,col)
        if row == @exit_row && col == @exit_col
          @labyrinth[row][col] = @@EXIT_CHAR
        elsif @monsters[row][col] != nil
          @labyrinth[row][col] = @@MONSTER_CHAR
        else
          @labyrinth[row][col] = @@EMPTY_CHAR

        end
      end
    end

    def dir_2_pos(row, col, direction)
      new_pos = [row, col]

      case direction
      when Directions::UP
        new_pos[@@ROW] -= 1
      when Directions::DOWN
        new_pos[@@ROW] += 1
      when Directions::LEFT
        new_pos[@@COL] -= 1
      when Directions::RIGHT
        new_pos[@@COL] += 1
      end

      new_pos
    end

    def random_empty_pos
      while true
        row = Dice.random_pos(@n_rows)
        col = Dice.random_pos(@n_cols)
        if empty_pos?(row,col)
          return new_pos = [row,col]
        end
      end
    end







  end
end